newlib/libc/sys/vita/dirent.c

```diff
@@ -24,74 +24,93 @@ DEALINGS IN THE SOFTWARE.
 
 #include <sys/dirent.h>
 #include <errno.h>
+#include <fcntl.h>
 #include <malloc.h>
+#include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
 
 #include <psp2/types.h>
 #include <psp2/io/dirent.h>
 #include <psp2/kernel/threadmgr.h>
-
-#define SCE_ERRNO_MASK 0xFF
-#define MAXNAMLEN	256
+#include "vitadescriptor.h"
+#include "vitaerror.h"
 
 struct DIR_
 {
-	SceUID uid;
+	int fd;
 	struct dirent dir;
-	char dirname[MAXNAMLEN];
 	int index;
 };
 
 int	closedir(DIR *dirp)
 {
-	if (!dirp || dirp->uid < 0)
+	if (!dirp || !dirp->fd)
 	{
 		errno = EBADF;
 		return -1;
 	}
 
-	int res = sceIoDclose(dirp->uid);
-	dirp->uid = -1;
-
+	int res = close(dirp->fd);
 	free(dirp);
 
 	if (res < 0)
 	{
-		errno = res & SCE_ERRNO_MASK;
 		return -1;
 	}
 
 	errno = 0;
 	return 0;
 }
 
-DIR *opendir(const char *dirname)
+DIR *__opendir_common(int fd)
 {
-	SceUID uid = sceIoDopen(dirname);
-
-	if (uid < 0)
-	{
-		errno = uid & SCE_ERRNO_MASK;
-		return NULL;
-	}
-
 	DIR *dirp = calloc(1, sizeof(DIR));
 
 	if (!dirp)
 	{
-		sceIoDclose(uid);
+		close(fd);
```

Comment:
is this right?
when call it from `fdopendir`, this chain would close the fd in here.

how about change the opendir to;
```
DIR *opendir(const char *dirname)
{
  int fd;
  if ((fd = ...) == -1)
    return NULL;
  DIR *ret = __opendir_common(fd);
  if (!ret) {
    close(fd);
  }
  return ret;
}
```

Reply:
Yeah. Nice catch.